internal/controller/apiextensions/composite/composition_functions.go

```diff
@@ -418,6 +425,29 @@ func (c *FunctionComposer) Compose(ctx context.Context, xr *composite.Unstructur
 	return CompositionResult{ConnectionDetails: d.GetComposite().GetConnectionDetails(), Composed: resources, Events: events}, nil
 }
 
+// The function `getComposedFieldOwnerName` generates a unique field owner name
+// for a given Crossplane composite resource (XR). This uniqueness is crucial to
+// prevent multiple XRs, which compose the same resource, from continuously
+// alternating as controllers.
+//
+// The function generates a deterministic hash based on the XR's name and
+// GroupKind (GK), ensuring consistency even during system restores. The hash
+// does not include the XR's UID (as it's not deterministic), namespace (XRs
+// don't have one), or version (to allow version changes without needing to
+// update the field owner name).
```

Comment:
If the XR is deleted and another is created with the same name, what is the desired behavior?  Is it guaranteed to be owned by the same field manager or is the new object lifecycle possibly owned by a different field manager?

Reply:
yeah, that would be the worst scenario, ideally we would like to have the old resource being deleted and a new one created, but if the deletion happens in background that is not guaranteed and if a new resource with the same name is created in the meantime, it could become the owner of the resource and happily switch the ownerRef without complaining because the field manager is the same. Which is exactly what happens today already. We would need to use a unique name for the specific XR to avoid that, right?